tools/syz-trace2syz/proggen/proggen.go

```diff
@@ -0,0 +1,613 @@
+}
+
+func genPtr(syzType *prog.PtrType, traceType parser.IrType, ctx *Context) prog.Arg {
+	switch a := traceType.(type) {
+	case *parser.PointerType:
+		if a.IsNull() {
+			return prog.DefaultArg(syzType)
+		}
+		if a.Res == nil {
+			res := genDefaultArg(syzType.Type, ctx)
+			return addr(ctx, syzType, res.Size(), res)
+		}
+		res := genArgs(syzType.Type, a.Res, ctx)
+		return addr(ctx, syzType, res.Size(), res)
+
+	case parser.Expression:
+		// Likely have a type of the form bind(3, 0xfffffffff, [3]);
+		res := genDefaultArg(syzType.Type, ctx)
+		return addr(ctx, syzType, res.Size(), res)
+	case *parser.Field:
+		return genPtr(syzType, a.Val, ctx)
+	case *parser.GroupType:
+		res := genArgs(syzType.Type, a, ctx)
+		return addr(ctx, syzType, res.Size(), res)
+	default:
+		res := genArgs(syzType.Type, a, ctx)
+		return addr(ctx, syzType, res.Size(), res)
+	}
+}
+
+func genConst(syzType prog.Type, traceType parser.IrType, ctx *Context) prog.Arg {
+	if syzType.Dir() == prog.DirOut {
+		return prog.DefaultArg(syzType)
+	}
+	switch a := traceType.(type) {
+	case parser.Expression:
+		switch b := a.(type) {
+		case parser.Ints:
+			if len(b) >= 2 {
+				// May get here through select. E.g. select(2, [6, 7], ..) since Expression can
+				// be Ints. However, creating fd set is hard and we let default arg through
+				return genDefaultArg(syzType, ctx)
+			}
+		}
+
+		return prog.MakeConstArg(syzType, a.Eval(ctx.Target))
+	case *parser.DynamicType:
+		return prog.MakeConstArg(syzType, a.BeforeCall.Eval(ctx.Target))
+	case *parser.GroupType:
+		// Sometimes strace represents a pointer to int as [0] which gets parsed
+		// as Array([0], len=1). A good example is ioctl(3, FIONBIO, [1]). We may also have an union int type that
+		// is a represented as a struct in strace e.g.
+		// sigev_value={sival_int=-2123636944, sival_ptr=0x7ffd816bdf30}
+		// For now we choose the first option
+		if a.Len == 0 {
+			log.Logf(2, "Parsing const type. Got array type with len 0")
+			return genDefaultArg(syzType, ctx)
+		}
+		return genConst(syzType, a.Elems[0], ctx)
+
+	case *parser.Field:
+		// We have an argument of the form sin_port=IntType(0)
+		return genArgs(syzType, a.Val, ctx)
+	case *parser.BufferType:
+		// The call almost certainly an error or missing fields
+		return genDefaultArg(syzType, ctx)
+		// E.g. ltp_bind01 two arguments are empty and
+	case *parser.PointerType:
+		// This can be triggered by the following:
+		// 2435  connect(3, {sa_family=0x2f ,..., 16)
+		return prog.MakeConstArg(syzType, a.Address)
+	default:
+		log.Fatalf("Cannot convert Strace Type: %s to Const Type", traceType.Name())
+	}
+	return nil
+}
+
+func genResource(syzType *prog.ResourceType, traceType parser.IrType, ctx *Context) prog.Arg {
+	if syzType.Dir() == prog.DirOut {
+		log.Logf(2, "Resource returned by call argument: %s", traceType.String())
+		res := prog.MakeResultArg(syzType, nil, syzType.Default())
+		ctx.ReturnCache.cache(syzType, traceType, res)
```

Comment:
Same here. Why do we want to cache the default value?

Reply:
For calls like pipe() which return file descriptors. I need to cache their value to track downstream usages.

Reply:
So this `traceType` is not actually a _type_, but rather a _value_?

Reply:
yeah i guess it is a type + value